File: TTH.c
#include<stdio.h>
#include<stdlib.h>
#include<math.h>
#include<string.h>
char alpha[26]={'a','b','c','d','e','f','g','h','i','j','k','l','m','n','o','p','q','r','s','t','u','v','w','x','y','z'};
int beta[26]={0,1,2,3,4,5,6,7,8,9,10,11,12,13,14,15,16,17,18,19,20,21,22,23,24,25};
char data[4][4],msg[40];
int fh[4],th[4],round2[4][4];
int main(){
	int len,m,i=1,j=0,k=0,l;
	
	printf("\n Toy Tetra Graph Hash");
	printf("\n-----------------------");
	printf("\n enter data(ignoring spaces,puntuatins):");
	scanf("%s",&msg);
	len=strlen(msg);
	if(len<16){
		for(m=len;m<16;m++){
			for(i=0;i<4;i++){
				for(j=0;j<4;j++) {
					data[i][j]=msg[m++]='\0';
				}
			}
		}
	}
	for(i=0;i<4;i++){
		for(j=0;j<4;j++){
			data[i][j]=msg[k];
			k++;
			if(k>strlen(msg)){
				break;
			}
		}
	}
	printf("\n data is: \n");
	for(i=0;i<4;i++){
		for(j=0;j<4;j++){
			printf("%c \t",data[i][j]);
		}
		printf("\n");
	}
	 i=0,j=0,k=0;
	char key;
	for(i=0;i<4;i++){
	for(j=0;j<4;j++){
	key=data[i][j];
	for(k=0;alpha[k]!='\0';k++){
	if(key==alpha[k]){
	data[i][j]=k;
	}
	}
	}
	th[0]=((data[0][0]+data[1][0]+data[2][0]+data[3][0])%26);
	th[1]=((data[0][1]+data[1][1]+data[2][1]+data[3][1])%26);
	th[2]=((data[0][2]+data[1][2]+data[2][2]+data[3][2])%26);
	th[3]=((data[0][3]+data[1][3]+data[2][3]+data[3][3])%26);
	printf("\n values of data hash are: \n");
	for(i=0;i<4;i++){
		printf("%d \t",th[i]);
	}
	//round 2 is matrix shifting of round 1
	round2[0][0]=data[0][1];
	round2[0][1]=data[0][2];
	round2[0][2]=data[0][3];
	round2[0][3]=data[0][0];
	round2[1][0]=data[1][2];
	round2[1][1]=data[1][3];
	round2[1][2]=data[1][0];
	round2[1][3]=data[1][1];
	round2[2][0]=data[2][3];
	round2[2][1]=data[2][0];
	round2[2][2]=data[2][1];
	round2[2][3]=data[2][2];
	round2[3][0]=data[3][3];
	round2[3][1]=data[3][2];
	round2[3][2]=data[3][1];
	round2[3][3]=data[3][0];
	//hash values of round2
	fh[0]=((round2[0][0]+round2[1][0]+round2[2][0]+round2[3][0])%26);
	fh[1]=((round2[0][1]+round2[1][1]+round2[2][1]+round2[3][1])%26);
	fh[2]=((round2[0][2]+round2[1][2]+round2[2][2]+round2[3][2])%26);
	fh[3]=((round2[0][3]+round2[1][3]+round2[2][3]+round2[3][3])%26);
	printf("\n values of round2 are: \n");
	for(i=0;i<4;i++){
	printf("%d \t",fh[i]);
	}
	
	i=0;
	printf("\n hash value is:");
	do{
	fh[i]=(fh[i]+th[i])%26;
	printf("\t %c",alpha[fh[i]]);
	i++;
	}while(i<4);
	}
	}

File: tth.py

import numpy as np

#assigning values
values={0: 'a', 1: 'b', 2: 'c', 3: 'd', 4: 'e', 5: 'f', 6: 'g', 7: 'h', 8: 'i', 9: 'j', 10: 'k', 11: 'l', 12: 'm', 13: 'n', 14: 'o', 15: 'p', 16: 'q', 17: 'r', 18: 's', 19: 't', 20: 'u', 21: 'v', 22: 'w', 23: 'x', 24: 'y', 25: 'z'}
values1={'a': 0, 'b': 1, 'c': 2, 'd': 3, 'e': 4, 'f': 5, 'g': 6, 'h': 7, 'i': 8, 'j': 9, 'k': 10, 'l': 11, 'm': 12, 'n': 13, 'o': 14, 'p': 15, 'q': 16, 'r': 17, 's': 18, 't': 19, 'u': 20, 'v': 21, 'w': 22, 'x': 23, 'y': 24, 'z': 25 }

print("\n Toy Tetra Graph Hash")
print("-------------------------")
msg=input("\nenter the message  : ")   
l=len(msg)
if(l<16):
	for i in range(l,16):
		msg= msg + 'a'


arr = np.array([[]])
for i in msg:
	arr = np.append(arr,values1[i])

print("\nThe message is : ")

arr = arr.reshape(4,4)
print("\n",arr)                                                             
                                                                       #round1
t1=arr.sum(axis=0)%26
print("\nthe values after round1 are :")                                                      
print("\n",t1)
                                                                       #round2
a1 = a2 = a3= arr
a1 = np.roll(a1[0],3)
a2 = np.roll(a2[1],2)
a3 = np.roll(a3[2],5)
a4 = np.fliplr(arr)
a4 = np.roll(a4[3],4)

t2 = (a1+a2+a3+a4)%26
print("\nthe values after round2 are :")
print("\n",t2)

tn=(t1+t2)%26
t = np.array([])
for i in tn:
	t=np.append(t,values[i])

print("\nthe hash value is :")
print("\n",t)




